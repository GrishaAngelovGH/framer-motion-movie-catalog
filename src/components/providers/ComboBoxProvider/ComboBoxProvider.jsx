import React from "react"

const comboBoxes = {
  "12wer": { id: "12wer", title: "Combo Box #1", products: [{ image: "/images/combo-coca-cola.jpg", title: "Coca-Cola", price: 3 }, { image: "/images/combo-chips.jpg", title: "Chips", price: 7 }, { image: "/images/combo-ice-cream.jpg", title: "Ice-Cream", price: 10 }] },
  "23dfg": { id: "23dfg", title: "Combo Box #2", products: [{ image: "/images/combo-fanta.jpg", title: "Fanta", price: 3 }, { image: "/images/combo-pizza.jpg", title: "Pizza", price: 8 }, { image: "/images/combo-doughnut.jpg", title: "Doughnut", price: 4 }] },
  "34sjd": { id: "34sjd", title: "Combo Box #3", products: [{ image: "/images/combo-7up.jpg", title: "7Up", price: 3 }, { image: "/images/combo-popcorn.jpg", title: "Popcorn", price: 5 }, { image: "/images/combo-kitkat.jpg", title: "Kit Kat", price: 4 }] }
}

const ComboBoxProvider = ({ children }) => React.cloneElement(children, { comboBoxes })

export default ComboBoxProvider