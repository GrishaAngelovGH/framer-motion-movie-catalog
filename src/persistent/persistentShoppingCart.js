const persistentShoppingCart = {
  add: function (id) {
    const shoppingCart = this.getShoppingCart()

    !shoppingCart[id] ? shoppingCart[id] = { qty: 1 } : shoppingCart[id].qty++

    window.localStorage.setItem("shoppingCart", JSON.stringify(shoppingCart))
  },
  hasItems: function () {
    const shoppingCart = this.getShoppingCart()
    return Object.keys(shoppingCart).length > 0
  },
  getShoppingCart: function () {
    return JSON.parse(window.localStorage.getItem("shoppingCart") || "{}")
  },
  updateQuantity: function (id, isIncremented) {
    const shoppingCart = this.getShoppingCart()
    isIncremented ? shoppingCart[id].qty++ : shoppingCart[id].qty > 1 && shoppingCart[id].qty--
    window.localStorage.setItem("shoppingCart", JSON.stringify(shoppingCart))
  },
  remove: function (id) {
    const shoppingCart = this.getShoppingCart()
    const { [id]: value, ...newShoppingCart } = shoppingCart
    window.localStorage.setItem("shoppingCart", JSON.stringify(newShoppingCart))
  }
}

export default persistentShoppingCart