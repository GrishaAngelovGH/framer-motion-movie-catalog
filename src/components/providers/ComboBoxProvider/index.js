import ComboBoxProvider from "./ComboBoxProvider"

export default ComboBoxProvider