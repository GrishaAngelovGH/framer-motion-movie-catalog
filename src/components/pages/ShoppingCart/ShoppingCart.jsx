import { useState } from "react"
import { Link } from "react-router-dom"

import persistentShoppingCart from "persistent/persistentShoppingCart"

import ShoppingCartItem from "./ShoppingCartItem"

const getShoppingCartItems = comboBoxes => {
  const shoppingCart = persistentShoppingCart.getShoppingCart()

  const shoppingCartItems = Object.entries(shoppingCart).map(([id, { qty }]) => {
    const comboPrice = comboBoxes[id].products.reduce((a, b) => a + b.price, 0)

    return ({
      id,
      qty,
      comboPrice,
      totalPrice: qty * comboPrice,
      ...comboBoxes[id]
    })
  })

  return shoppingCartItems
}

const ShoppingCart = ({ comboBoxes }) => {
  const [shoppingCartItems, setShoppingCartItems] = useState(getShoppingCartItems(comboBoxes))

  const handleRemoveShoppintCartItem = id => {
    persistentShoppingCart.remove(id)
    setShoppingCartItems(getShoppingCartItems(comboBoxes))
  }

  return (
    <div className="h-screen bg-gray-100">
      <div className="bg-blue-400 text-center text-white p-2 flex justify-center items-center">
        <Link to="/experiences" className="text-white hover:text-white absolute left-3">
          <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" strokeWidth={1.5} stroke="currentColor" className="size-6">
            <path strokeLinecap="round" strokeLinejoin="round" d="M9 15 3 9m0 0 6-6M3 9h12a6 6 0 0 1 0 12h-3" />
          </svg>
        </Link>

        <div className="text-4xl">Shopping Cart</div>
      </div>
      <div className="mt-20 flex flex-col md:flex-row justify-center gap-5">
        {
          shoppingCartItems.map(v => (
            <ShoppingCartItem
              key={v.id}
              {...v}
              onRemove={handleRemoveShoppintCartItem}
            />
          ))
        }
      </div>
    </div>
  )
}

export default ShoppingCart