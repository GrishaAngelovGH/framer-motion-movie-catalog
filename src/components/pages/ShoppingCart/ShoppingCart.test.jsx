import { render } from "@testing-library/react"
import { BrowserRouter as Router, Routes, Route } from "react-router-dom"
import { vi } from "vitest"

import ShoppingCart from "./ShoppingCart"
import ComboBoxProvider from "components/ComboBoxProvider"

import persistentShoppingCart from "persistent/persistentShoppingCart"

test("should render pages/ShoppingCart component", () => {
  vi.spyOn(persistentShoppingCart, "getShoppingCart").mockReturnValue({ "12wer": { qty: 2 } })

  const view = render(
    <Router>
      <Routes>
        <Route path="/" element={
          <ComboBoxProvider>
            <ShoppingCart />
          </ComboBoxProvider>
        } />
      </Routes>
    </Router>
  )

  expect(view).toMatchSnapshot()
})