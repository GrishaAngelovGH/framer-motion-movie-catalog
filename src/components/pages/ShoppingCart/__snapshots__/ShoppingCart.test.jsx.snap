// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render pages/ShoppingCart component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-screen bg-gray-100"
      >
        <div
          class="bg-blue-400 text-center text-white p-2 flex justify-center items-center"
        >
          <a
            class="text-white hover:text-white absolute left-3"
            href="/experiences"
          >
            <svg
              class="size-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 15 3 9m0 0 6-6M3 9h12a6 6 0 0 1 0 12h-3"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </a>
          <div
            class="text-4xl"
          >
            Shopping Cart
          </div>
        </div>
        <div
          class="p-10 flex flex-col md:flex-row justify-center gap-5"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-screen bg-gray-100"
    >
      <div
        class="bg-blue-400 text-center text-white p-2 flex justify-center items-center"
      >
        <a
          class="text-white hover:text-white absolute left-3"
          href="/experiences"
        >
          <svg
            class="size-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 15 3 9m0 0 6-6M3 9h12a6 6 0 0 1 0 12h-3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </a>
        <div
          class="text-4xl"
        >
          Shopping Cart
        </div>
      </div>
      <div
        class="p-10 flex flex-col md:flex-row justify-center gap-5"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
